### PR TITLE
Feature: set OpenMP thread num to a proper default value

### LINKED ISSUE #3701: Feature: set default OpenMP parallel number to 1
### Background

For those ABACUS freshmen, when they just use `mpirun -np $NPROCS abacus` to run abacus, they tend to forget about the OpenMP parallel setting OMP_NUM_THREADS, for which in VASP is 1 and often only 1, but sometimes in ABACUS will be $NPROC and lead to low performance.

### Describe the solution you'd like

Set default OMP_NUM_THREADS=1 in ABACUS itself will be a user-friendly option, for users will usually use MPI first.

### Task list only for developers

- [X] Notice possible changes of behavior
- [ ] Explain the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi

### Notice Possible Changes of Behavior (Reminder only for developers)

_No response_

### Notice any changes of core modules (Reminder only for developers)

_No response_

### Notice Possible Changes of Core Modules (Reminder only for developers)

_No response_

### Additional Context

_No response_

### Task list for Issue attackers (only for developers)

- [ ] Review and understand the proposed feature and its importance.
- [ ] Research on the existing solutions and relevant research articles/resources.
- [ ] Discuss with the team to evaluate the feasibility of implementing the feature.
- [ ] Create a design document outlining the proposed solution and implementation details.
- [ ] Get feedback from the team on the design document.
- [ ] Develop the feature following the agreed design.
- [ ] Write unit tests and integration tests for the feature.
- [ ] Update the documentation to include the new feature.
- [ ] Perform code review and address any issues.
- [ ] Merge the feature into the main branch.
- [ ] Monitor for any issues or bugs reported by users after the feature is released.
- [ ] Address any issues or bugs reported by users and continuously improve the feature.

### PR DESCRIPTION
**Pull request by caic99**
### Reminder
- [ ] Have you linked an issue with this pull request?
- [ ] Have you added adequate unit tests and/or case tests for your pull request?
- [ ] Have you noticed possible changes of behavior below or in the linked issue?
- [ ] Have you explained the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi? (ignore if not applicable)

### Linked Issue
Fix #3701 

### Unit Tests and/or Case Tests for my changes
- A unit test is added for each new feature or bug fix.

### What's changed?
ABACUS will try to set a proper OpenMP threading value if `OMP_NUM_THREADS` is not set, and exit with error if the value set is beyond hardware availability (which should not happen since the value has been automatically set).

### CODE REVIEW

**Comment by caic99** [comment]
Could someone please help me on [this error](https://github.com/deepmodeling/abacus-develop/actions/runs/8625588330/job/23642411212?pr=3943#step:5:948) building tests?

**Comment by QuantumMisaka** [comment]
@caic99 I have some test by using this repo. It worked well , but if the machine has hyper-thread (like WSL or bohrium image), the product of OMP number and MPI number will include hyper-thread and lead to lower parallel efficiency than the hardware-core (hyper-thread / 2) situation.

**Comment by caic99** [comment]
> @caic99 I have some test by using this repo. It worked well , but if the machine has hyper-thread (like WSL or bohrium image), the product of OMP number and MPI number will include hyper-thread and lead to lower parallel efficiency than the hardware-core (hyper-thread / 2) situation.

It's hard to determine with standard c++ API whether hyperthreading is used.

**Comment by mohanchen** [comment]
Any updates?

**Comment by caic99** [comment]
> Could someone please help me on [this error](https://github.com/deepmodeling/abacus-develop/actions/runs/8625588330/job/23642411212?pr=3943#step:5:948) building tests?

@mohanchen help needed 😢

File: source/module_base/parallel_global.cpp [hunk 1]
```diff
@@ -176,39 +176,44 @@ void Parallel_Global::read_mpi_parameters(int argc,char **argv)
 	// get the rank --> GlobalV::MY_RANK
 	MPI_Comm_size(MPI_COMM_WORLD,&GlobalV::NPROC);
     MPI_Comm_rank(MPI_COMM_WORLD, &GlobalV::MY_RANK);
+    MPI_Comm shmcomm;
+    MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_SHARED, 0, MPI_INFO_NULL, &shmcomm);
```
Comment (mohanchen) [inline source/module_base/parallel_global.cpp:182]: add some explanations?

Comment (caic99) [reply source/module_base/parallel_global.cpp:182]: @mohanchen Sure, would you see the latest commit.

File: source/module_base/parallel_global.cpp [hunk 2]
```diff
@@ -172,43 +172,56 @@ void Parallel_Global::read_mpi_parameters(int argc,char **argv)
 
     //  GlobalV::KPAR = atoi(argv[1]); // mohan abandon 2010-06-09
 
-	// get the size --> GlobalV::NPROC
-	// get the rank --> GlobalV::MY_RANK
-	MPI_Comm_size(MPI_COMM_WORLD,&GlobalV::NPROC);
+    // get world size --> GlobalV::NPROC
```
Comment (mohanchen) [inline source/module_base/parallel_global.cpp:175]: Could you use the format of Doxygen? https://mcresearch.github.io/abacus-user-guide/develop-dox.html

Comment (caic99) [reply source/module_base/parallel_global.cpp:175]: This is a part of comments, and doxygen will not generate documentation for this part. It's adequate to add the comment to the function/variable declaration in the header file, i.e. global_variable.h.

**Review by mohanchen — APPROVED** [review]

**Pull request merged**